### ISSUE TITLE
MultibodyPlant vs Pinocchio Performance

### ISSUE THREAD

**Issue by mposa** [DAIRLab/dairlib#181]
This is an issue to track Drake::MBP vs alternatives (currently Pinocchio). This example is on a Cassie URDF. Some comments on the plant:

- Does not include loop closure
- Does not include contact
- Fixed base model (pelvis)

To replicate:

- Clone dairlib
- Checkout `pinocchio` branch of dairlib
- Install Pinocchio (https://github.com/stack-of-tasks/pinocchio). Install directions at https://stack-of-tasks.github.io/pinocchio/download.html. You will _not_ need to configure environment variables
- Build and run `examples/Cassie:benchmark_dynamics`. For non-Penn users, recommend
```
bazel run --define WITH_SNOPT=OFF examples/Cassie:benchmark_dynamics 
```

## Status, July 18, 2020
As run on Michael's laptop
```
(multibody_plant) 100x inertia autodiff calculations took 182 miliseconds. 1820 microseconds per.
(multibody_plant) 10000x inverse dynamics calculations took 125 miliseconds. 12 microseconds per.
(multibody_plant) 100xautodiff inverse dynamics calculations took 240 miliseconds. 2400 microseconds per.
(multibody_plant) 10000x forward dynamics calculations took 304 miliseconds. 30 microseconds per.
(multibody_plant) 100xautodiff forward dynamics calculations took 348 miliseconds. 3480 microseconds per.
(pinocchio) 10000x rnea inverse dynamics calculations took 20 miliseconds. 2 microseconds per.
(pinocchio) 10000x rnea inverse dynamics derivatives calculations took 61 miliseconds. 6 microseconds per.
(pinocchio) 10000x aba forward dynamics calculations took 32 miliseconds. 3 microseconds per.
(pinocchio) 10000x aba forward dynamics derivative calculations took 120 miliseconds. 12 microseconds per.
vdot_mbp, vdot_pin, diff
```

### Summary (Drake, Pinocchio)
For numerical derivatives, approximating by double x 42 (counting fixed-base coordinates `nq=nv=16` and `nu=10` )
- Forward dynamics: (30, 3)
- Inverse dynamics: (12, 3)
- Forward dynamics derivatives , Drake-Autodiff: (3480, 12)
- Forward dynamics derivatives , Drake-Numerical: (1260, 12)
- Inverse dynamics derivatives , Drake-Autodiff: (2400, 6)
- Inverse dynamics derivatives , Drake-Numerical: (504, 6)

**Comment by sherm1** [comment]
I'm making some progress on this -- found some inadvertent inner-loop heap allocations in MBP. Not sure yet how much those affect performance but they can't be good. Will fix shortly.

**Comment by sherm1** [comment]
FYI @mposa there appear to be some bugs in the test cases for autodiff of inverse & forward dynamics. Try setting the iteration counts to 1 in the test and run a Debug configuration (so that all the Eigen asserts are enabled). Looks like inconsistent use if `drake::math::initializeAutoDiff()` producing different numbers of derivatives for (e.g.) x and u in forward dynamics. Possible fixes mentioned [here](https://github.com/RobotLocomotion/drake/pull/13782#pullrequestreview-459532490).

**Comment by mposa** [comment]
Yes, you're right. I caught that when I added the comparison between gradient matrices, but did not bother fixing the for loops. The bug probably makes AutoDiff actually look better than it really is.

**Comment by yminchen** [comment]
A few benchmark regard the center of mass:

All units below are [seconds per 1,000,000 calls].

Without resetting drake context every loop:
r_com time (Pinocchio):0.833135
r_com time (Drake):0.63641
v_com time (Pinocchio):1.38537
v_com time (Drake):1.50729

When resetting context in every loop:
r_com time (Pinocchio):1.09758
r_com time (Drake):3.59607
v_com time (Pinocchio):1.63794
v_com time (Drake):6.92368
J_com time (Pinocchio):1.29014
J_com time (Drake):13.6425

**Comment by sherm1** [comment]
Thanks @yminchen! I see that the Pinocchio times changed in the second set also -- is there an equivalent of resetting the Context in Pinocchio? Also, can you say what you did specifically to "reset" the Context in Drake?

**Comment by sherm1** [comment]
cc @amcastro-tri

**Comment by yminchen** [comment]
@sherm1 Thanks for the question! I should've mentioned this in the previous post. The benchmark was done not directly on Pinocchio but PinocchioPlant (we derived this class from Drake's MultibodyPlant by overriding center of mass related methods). This is why there is a difference between with and without resetting context for Pinocchio. The extra runtime is from calling `plant.SetPositions(context.get(), q)`, for example.
